### PR TITLE
Issue 506 integration test

### PR DESCRIPTION
**Pull request by JennyLouise**
### Issue reference / description

The branch you merge from should already reference an event-formation-unit github ticket number. You can add a descriptive title, but if an issue is referenced, you don't have to.

## Checklist for submitter

- [ ] Check for conflict with integration test
- [ ] Unit tests pass

---

## Nominate for Group Code Review (Anyone can nominate it)
Indicate if you think the code should be reviewed in a Thursday code review session.

- [ ] Recommend for group code review

Also, nominate it on the code_review Slack channel.

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -306,3 +311,18 @@ timestamps {
...
+    // This is a pull request build
+    node('inttest') {
+        stage('Integration Test') {
+            checkout scm
+            unstash 'event-formation-unit-centos7.tar.gz'
+            sh "tar xzvf event-formation-unit-centos7.tar.gz"
+            sh "ls -R"
+            sh """
+                export LD_LIBRARY_PATH=\$LD_LIBRARY_PATH:./event-formation-unit/lib/
+                python3 -u ./utils/efushell/integrationtest.py
```
Comment (mortenjc) [inline Jenkinsfile:324]: The integration test is well hidden here. It is also less a utility and more athing in its own right. Maybe put this in its own folder at the top level so it is easy to spot (or in its ownfolder below utils).

Also consider updating README.md

File: utils/efushell/integrationtest.json [hunk 2]
```diff
@@ -0,0 +1,16 @@
+
```
Comment (mortenjc) [inline utils/efushell/integrationtest.json:1]: Using tabs for indent or four spaces (I can't tell)? Let's stick to spaces. I normally use 2 spaces just as we do for C++ code. Might be worth doing consistently for this repo?

File: utils/efushell/integrationtest.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+from EFUMetrics import Metrics
+
+
+def compare_pair(x, y, operator):
+    if operator == "==":
+        return x == y
+    if operator == ">=":
+        return x >= y
+    print("invalid operator")
+    return False
```
Comment (mortenjc) [inline utils/efushell/integrationtest.py:13]: maybe it is better to raise an exception here to be able to distinguish between the check and the valid command?

File: utils/efushell/integrationtest.json [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+{
+"Tests":	
+	[
+		{
+			"Module":"loki",
+			"Config":"./src/modules/loki/configs/STFCTestII.json",
+			"KafkaTopic":"loki_detector",
+			"Generator" : "loki_udp_generated",
+			"Packets" : 50,
```
Comment (mortenjc) [inline utils/efushell/integrationtest.json:10]: The test should check for the exact number of packets to be received by the efu.

File: utils/efushell/integrationtest.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+                        messages += result['count']
+        if messages == test["ExpectedKafkaMessages"]:
+                print(f"Successfully received {test["ExpectedKafkaMessages"]} messages in Kafka topic {test["KafkaTopic"]}")
+        else:
+                raise Exception(f"Did not successfully receive {test["ExpectedKafkaMessages"]} messages in Kafka topic {test["KafkaTopic"]}, received {messages}")
+
+def run_tests():
+        efu = "./event-formation-unit"
+        stats_test_list = [
+                                                ['kafka.ev_errors', '==', 0],
```
Comment (mortenjc) [inline utils/efushell/integrationtest.py:88]: add: receive.packets ?

**Review by mortenjc — APPROVED** [review]

File: utils/efushell/integrationtest.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+	return efu_process
+
+
+def run_data_generator(test, efu):
+	print("Running Data Generator")
+	generator_process = subprocess.Popen(f"{efu}/generators/{test['Generator']} -a {test['Packets']} -o 50 -t {test['Throttle']}", shell=True)
+	exit_code = generator_process.wait()
+	if exit_code != 0:
+		raise Exception(f"Artififial generator {test['Generator']} failed")
+	time.sleep(5)
```
Comment (JennyLouise) [inline utils/efushell/integrationtest.py:37]: remove sleep

File: utils/efushell/integrationtest.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+	return generator_process
+
+
+def check_stats(test, stats_test_list):
+	print("Getting EFU Stats")
+	
+	for i in range(10):
```
Comment (JennyLouise) [inline utils/efushell/integrationtest.py:44]: Get rid of loop

File: utils/efushell/integrationtest.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+		else:
+			print(f"Stat check passed for {stats_test[0]} {stats_test[1]} {stats_test[2]}")
+
+
+
+def create_kafka_topic(topic_name):
+        subprocess.Popen(f"/ess/ecdc/kafka/kafka_2.13-2.8.0/bin/kafka-topics.sh --bootstrap-server localhost:9092 --create --topic {topic_name}", shell=True).wait()
+
+def check_kafka(test):
+        kafka_process = subprocess.Popen(f"/ess/ecdc/kafka/kafka_2.13-2.8.0/bin/kafka-verifiable-consumer.sh --bootstrap-server localhost:9092 --topic {test["KafkaTopic"]} --group-id testconsumer1", shell=True, stdout=subprocess.PIPE)
```
Comment (amues) [inline utils/efushell/integrationtest.py:71]: The double quotes for the dictionary keys are causing a syntax error.

File: utils/efushell/integrationtest.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+        time.sleep(5)
+        kafka_process.kill()
+        out, err = kafka_process.communicate()
+        results_dict = [json.loads(str(x)) for x in out.decode("utf-8").split("\n") if str(x) != '']
+        messages = 0
+        for result in results_dict:
+                if result['name'] == "records_consumed":
+                        messages += result['count']
+        if messages == test["ExpectedKafkaMessages"]:
+                print(f"Successfully received {test["ExpectedKafkaMessages"]} messages in Kafka topic {test["KafkaTopic"]}")
```
Comment (amues) [inline utils/efushell/integrationtest.py:81]: The double quotes for the dictionary keys are causing a syntax error.

File: utils/efushell/integrationtest.py [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+        else:
+                raise Exception(f"Did not successfully receive {test["ExpectedKafkaMessages"]} messages in Kafka topic {test["KafkaTopic"]}, received {messages}")
```
Comment (amues) [inline utils/efushell/integrationtest.py:83]: The double quotes for the dictionary keys are causing a syntax error.

File: utils/efushell/integrationtest.py [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+def run_tests():
+        efu = "./event-formation-unit"
+        stats_test_list = [
+                                                ['kafka.ev_errors', '==', 0],
+                                                ['transmit.bytes', '>=', 500000]
+        ]
+
+        file = open('./utils/efushell/integrationtest.json')
+        data = json.load(file)
+        file.close()
```
Comment (amues) [inline utils/efushell/integrationtest.py:94]: Here you could use a context manager that takes care of automatically closing the file after the operations using it:
```suggestion
        with open('./utils/efushell/integrationtest.json') as f:
            data = json.load(f)
```

**Review by amues — APPROVED** [review]
The code looks good, there is only an issue with the double quotes (see comments).

**Comment by amues** [comment]
A suggestion I have is to run `black` on the code for automated formatting (you can install it with `pip install black` and then run `black <filename>`). It's not strictly required, but we use it in some of our Python projects to standardise code formatting.

File: utils/efushell/integrationtest.py [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+			break
+		except:
+			print("failed to get metrics")
+
+	for stats_test in stats_test_list:
+		actual_stat = metrics.return_metric(f"efu.{test['Module']}.0.{stats_test[0]}")
+		operator = stats_test[1]
+		expected_stat = stats_test[2]
+		if actual_stat == -1:
+			efu_process.kill()
```
Comment (amues) [inline utils/efushell/integrationtest.py:57]: `efu_process` seems to be undefined here.

File: utils/efushell/integrationtest.py [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+			raise Exception(f"EFU stat check failed, did not find stat named {stats_test[0]}")
+		if not compare_pair(actual_stat, expected_stat, operator):
+			efu_process.kill()
```
Comment (amues) [inline utils/efushell/integrationtest.py:60]: `efu_process` seems to be undefined here as well.

**Pull request merged**